internal/driver/browser_test.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package driver
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/chromedp/chromedp"
+)
+
+func skipUnlessHaveChrome(t *testing.T) {
+	if _, err := exec.LookPath("google-chrome"); err == nil {
+		return
+	}
+	if _, err := exec.LookPath("chrome"); err == nil {
+		return
+	}
+	t.Skip("chrome not available")
+}
+
+func TestTopTable(t *testing.T) {
+	skipUnlessHaveChrome(t)
+
+	prof := makeFakeProfile()
+	server := makeTestServer(t, prof)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+	ctx, cancel = chromedp.NewContext(ctx)
+	defer cancel()
+
+	err := chromedp.Run(ctx,
+		chromedp.Navigate(server.URL+"/top"),
+		chromedp.WaitVisible(`#toptable`, chromedp.ByID),
+
+		// Check that fake profile entries show up in the right order.
+		matchRegexp(t, "#node0", `200ms.*F2`),
+		matchInOrder(t, "#toptable", "F2", "F3", "F1"),
+
+		// Check sorting by cumulative count.
+		chromedp.Click(`#cumhdr1`, chromedp.ByID),
+		matchInOrder(t, "#toptable", "F1", "F2", "F3"),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// matchRegexp is a chromedp.Action that fetches the text of the first
+// node that matched query and checks that the text matches regexp re.
+func matchRegexp(t *testing.T, query, re string) chromedp.ActionFunc {
+	return func(ctx context.Context) error {
+		var value string
+		err := chromedp.Text(query, &value, chromedp.ByQuery).Do(ctx)
+		if err != nil {
+			return fmt.Errorf("text %s: %v", query, err)
+		}
+		t.Logf("text %s:\n%s", query, value)
```

Comment:
I like leaving things like this in because if the test breaks, it is something useful for the fixer to look at -- they will see the extracted HTML.

A plain "go test" shows t.Logf() output only for failing runs. We should consider dropping the "-v" argument from the "go test" call in test.sh to reduce the clutter from our workflow runs.

Reply:
Sounds good, including removing "-v". I think having the verbose output was helpful a couple of times, but I agree in most cases it's more likely to get in the way rather than be useful (often takes a lot of scrolling to see the error message).